sqlitecluster/SQLiteNode.cpp

```diff
@@ -74,6 +80,143 @@ SQLiteNode::~SQLiteNode() {
+                commandCommitCount = command.calcU64("NewCount");
+                // Indicate that this command is starting.
+                transactionState = 1;
+            }
+            if (SIEquals(command.methodLine, "ROLLBACK_TRANSACTION")) {
+                lock_guard<mutex> hashlock(node._replicationHashMutex);
+                node._rollbackHashes.insert(command["NewHash"]);
+                node._replicationCV.notify_all();
+                continue;
+            }
+            if (SIEquals(command.methodLine, "COMMIT_TRANSACTION")) {
+                commandCommitCount = command.calcU64("CommitCount");
+                lock_guard<mutex> hashlock(node._replicationHashMutex);
+                node._replicationHashes.emplace(make_pair(commandCommitCount, command["Hash"]));
+                node._replicationCV.notify_all();
+                continue;
+            }
+        }
+
+        try {
+            // Convenience vars.
+            auto& peer = p.first;
+            auto& command = p.second;
+
+            // The following block only runs if there's already a command ready to run. This includes a command we just
+            // created
+            if (transactionState == 1) {
+                unique_lock<mutex> lock(node._replicationCommandMutex);
+                if (node._replicationThreadsShouldExit) {
+                    return;
+                }
+                if (commandCommitCount == db.getCommitCount() + 1) {
+                    // We can unlock once we know our condition has passed, there's no race in case it changes after
+                    // we've checked it but before we wait again, as we wont wait. We do this before any DB operations
+                    // so that waiting on the DB can't block enqueueing new commands. But it's important that we do
+                    // hold this lock before checking the conditions that determine if we can proceed with our DB
+                    // operations.
+                    lock.unlock();
+
+                    // Importantly, we don't start our transaction until the previous transaction has fully completed.
+                    // In the next version, where we handle concurrent transactions, this will need to be able to start
+                    // regardless of whether the previous transaction has started.
+                    node.handleBeginTransaction(db, peer, command);
+                    transactionState = 2;
+                } else {
+                    // Wait and then start from the beginning.
+                    node._replicationCV.wait(lock);
+                    continue;
+                }
+            }
+
+            // And the next step.
+            if (transactionState == 2) {
+                unique_lock<mutex> lock(node._replicationCommandMutex);
+                if (node._replicationThreadsShouldExit) {
+                    db.rollback();
+                    return;
+                }
+                // Look up our hashes.
```

Comment:
Please add a blank line above this.  Also, this comment is a bit terse, how about:
```
// Look to see if the transaction this thread began should be rolled back by checking to see if
// some other replication thread flagged this thread's hash for rollback in the `_rollbackHashes` list
```